### PR TITLE
Mithril relay P2P network for signatures PoC

### LINKED ISSUE #1300: P2P Signature Registration PoC
## Why
In order to move toward a decentralized Mithril network that would allow multiple aggregators to operate simultaneously.
A P2P network layer could power this new capability and provide a way to exchange signatures between signers and aggregators without relying on a central entity. 

The goal is to create a Proof of Concept for a basic P2P network that would allow multiple aggregator to retrieve the individual signatures posted by signers. A nice to have feature is the ability to deploy a P2P network and at the same time keep the existing centralized implementation: this would allow in the long run to smoothly deploy the P2P network along the existing network.

A good candidate to do so is the **Mithril relay** that we could re-implement:
- Mithril relay is a "proxy" to the historic aggregator.
- Mithril relay is also able to replicate individual signatures sent by the signer to a P2P network (each relay is a peer in the network).

Along the **Mithril relay**, the **Mithril aggregator** has an adapter that is able to retrieve the signatures from the P2P network and make them available to its Certificate producer. 

Here is a draft schema for the architecture of such a P2P **Mithril network**:
Mithril P2P Network

**Note**: It is worth mentioning that more work will be needed to work in fully a decentralized setup:
- Capability for an aggregator of peer-discovering other aggregators in the network.
- Capability for an aggregator of retrieving from another aggregator a valid certificate chain.
- Capability for a signer to compute message to sign without communicating with an aggregator.
- Capability for a signer to register its verification keys in a decentralized manner (**with consensus**).

A good P2P candidate library that we intend to evaluate is [libp2p](https://libp2p.io/):
- It has a Rust implementation that is mature and stable
- It provides support for interesting P2P features: **PubSub** [guide](https://docs.libp2p.io/concepts/pubsub/overview/) | [library](https://docs.rs/libp2p/latest/libp2p/gossipsub/)
- It has good documentation and [tutorials](https://docs.rs/libp2p/latest/libp2p/tutorials/ping/index.html)

## What
We want to experiment with P2P and create a PoC to decentralize the signature registration of the signers and make it available to subscribers on a P2P network. In that context, an aggregator subscribed to the signatures can create certificates by relying only on the P2P signatures broadcast. The selected architecture should be versatile enough to not modify the existing signer and aggregator nodes. 

## How

- [x] Make a working prototype of a relay with libp2p **PubSub** (#1326)
- [x] Prepare a dev testnet deployment (#1361)
- [x] Assess problems/difficulties (identified during the implementation phase)

### Later
- Work on other features: client and aggregator use peer discovery to select an aggregator, ...
- Assess what's missing to have multiple aggregators producing certificates

### LINKED ISSUE #1326: Prototype a P2P relay with `libp2p`
## Why
We want to experiment P2P networking with the `libp2p` library, and particularly with the `PubSub` decentralized queue that it offers.

## What
In order, to get acquainted with the technology and to better understand if `libp2p` is a good fit, we want to build a very basic **Mithril relay** that would only be in charge of publishing single signatures to a decentralized **PubSub** topic when being called on a REST API. P2P clients (aka aggregators) would subscribe to the topic and receive notifications with the incoming signatures after their publication.

## How

- [x] Implement basic `client` peer node in a P2P network that subscribes to a P2P pubsub queue and listens to incoming events
- [x] Implement basic `signer relay` peer node in a P2P network that exposes a REST API to send signatures and publishes to a topic of a P2P pubsub queue, and provides a gateway to required routes of the aggregator REST API
- [x] Implement basic `aggregator relay` peer in a P2P network that listens to signatures published on a topic of a P2P pubsub queue and forwards them to the existing REST API of the aggregator
- [x] Wire the relays in the end to end test under a configuration for activating the P2P network mode

## Issue(s)
Epic #1300

### PR DESCRIPTION
**Pull request by jpraynaud**
## Content
This PR includes the implementation of a new **Mithril relay** prototype node, that will in the long run supersede the current implementation.

This new node is a relay for the signer and the aggregator, which enables us to experiment with P2P networking with `libp2p` locally with the end to end test, and also later for testing purpose on the testing networks, without modification neither on the signer nor the aggregator nodes.

## Pre-submit checklist

- Branch
  - [x] Tests are provided (if possible)
  - [x] Crates versions are updated (if relevant)
  - [x] Commit sequence broadly makes sense
  - [x] Key commits have useful messages
- PR
  - [x] No clippy warnings in the CI
  - [x] Self-reviewed the diff
  - [x] Useful pull request description
  - [x] Reviewer requested
- Documentation
  - [x] Update README file (if relevant)

### To do later
- Fix CI flakiness due to P2P latency vs end to end test parameters
- Remove `ping` behavior in peer
- Make integration test with aggregator relay and signer relay (and cleanup P2P client)
- Deploy separate commands for node types in infrastructure
- Deploy network with P2P actionable feature (see [ensemble/1300-deploy-relay-p2p-network](https://github.com/input-output-hk/mithril/tree/ensemble/1300-deploy-relay-p2p-network))
- Make integration test with variable number of aggregator relays and signer relays
- Implement API version in the messages sent on the P2P network
- Make relay era aware (needs Cardano node access)?
- Replace Test HTTP server with production HTTP server
- Add OpenAPI specs tests (as in aggregator)
- Add optional usage of relay in `devnet`
- Add `AggregatorClient` layer

## Issue(s)
Closes #1326

### CODE REVIEW

File: mithril-relay/src/relay/signer.rs [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    }
+
+    /// Receive signature from the underlying channel
+    #[allow(dead_code)]
+    pub async fn receive_signature(&mut self) -> Option<RegisterSignatureMessage> {
+        self.signature_rx.recv().await
+    }
+
+    /// Tick the peer of the signer relay
+    /// #[allow(dead_code)]
```
Comment (dlachaumepalo) [inline mithril-relay/src/relay/signer.rs:101]: I think that this could be removed now.

**Review by dlachaumepalo — APPROVED** [review]
LGTM, just a comment on the `allow(dead_code)` attribute to remove and the typo to fix.

File: mithril-test-lab/mithril-end-to-end/src/mithril/infrastructure.rs [hunk 2]
```diff
@@ -48,15 +52,46 @@ impl MithrilInfrastructure {
         });
         aggregator.serve()?;
 
+        let mut relay_aggregators: Vec<RelayAggregator> = vec![];
+        let mut relay_signers: Vec<RelaySigner> = vec![];
+        if use_p2p_network_mode {
+            info!("Starting the Mithirl infrastructure in P2P mode (experimental)");
```
Comment (dlachaumepalo) [inline mithril-test-lab/mithril-end-to-end/src/mithril/infrastructure.rs:58]:
```suggestion
            info!("Starting the Mithril infrastructure in P2P mode (experimental)");
```

**Review by dlachaumepalo — COMMENTED** [review]
`mithril-end-to-end` crate version should be incremented.

File: mithril-test-lab/mithril-end-to-end/src/mithril/infrastructure.rs (hunk 2 above)
Comment (jpraynaud) [reply mithril-test-lab/mithril-end-to-end/src/mithril/infrastructure.rs:58]: Done :heavy_check_mark:

File: mithril-relay/src/relay/signer.rs (hunk 1 above)
Comment (jpraynaud) [reply mithril-relay/src/relay/signer.rs:101]: Done :heavy_check_mark:

File: mithril-relay/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,16 @@
+#![warn(missing_docs)]
+
+//! Mithril relay modules
+
+mod commands;
+mod p2p;
+mod relay;
+
+pub use commands::RelayCommands;
+pub use p2p::*;
```
Comment (ghubertpalo) [inline mithril-relay/src/lib.rs:10]: Maybe make the `p2p` module public.

File: mithril-relay/src/p2p/peer.rs [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    /// Publish a signature on the P2P pubsub
+    pub fn publish_signature(
+        &mut self,
+        message: &RegisterSignatureMessage,
+    ) -> StdResult<gossipsub::MessageId> {
+        let topic = self
+            .topics
+            .get(MITHRIL_SIGNATURES_TOPIC_NAME)
+            .ok_or(PeerError::MissingTopic())
+            .with_context(|| "Can not publish signature on invalid topic")?
```
Comment (ghubertpalo) [inline mithril-relay/src/p2p/peer.rs:180]: Is it possible to specify the topic name in the error message?

File: mithril-relay/src/p2p/peer.rs [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+            .with_context(|| "Can not publish signature on P2P pubsub")?
+            .ok_or(PeerError::UnavailableSwarm())
+            .with_context(|| "Can not publish signature without swarm")?;
+        Ok(message_id.to_owned())
+    }
+
+    /// Connect to a remote peer
+    pub fn dial(&mut self, addr: Multiaddr) -> StdResult<()> {
+        debug!("Peer: dialing to"; "address" => format!("{addr:?}"), "local_peer_id" => format!("{:?}", self.local_peer_id()));
+        self.swarm.as_mut().unwrap().dial(addr)?;
```
Comment (ghubertpalo) [inline mithril-relay/src/p2p/peer.rs:199]: this can be returned as result

File: mithril-relay/src/relay/aggregator.rs [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        }
+    }
+
+    async fn notify_signature_to_aggregator(
+        &self,
+        signature_message: &RegisterSignatureMessage,
+    ) -> StdResult<()> {
+        let response = reqwest::Client::new()
+            .post(format!("{}/register-signatures", self.aggregator_endpoint))
+            .json(signature_message)
```
Comment (ghubertpalo) [inline mithril-relay/src/relay/aggregator.rs:41]: maybe this code could be moved in an `AggregatorClient` layer?

**Review by ghubertpalo — APPROVED** [review]

File: mithril-relay/src/lib.rs (hunk 3 above)
Comment (jpraynaud) [reply mithril-relay/src/lib.rs:10]: Done :heavy_check_mark:

File: mithril-relay/src/p2p/peer.rs (hunk 4 above)
Comment (jpraynaud) [reply mithril-relay/src/p2p/peer.rs:180]: Done :heavy_check_mark:

File: mithril-relay/src/p2p/peer.rs (hunk 5 above)
Comment (jpraynaud) [reply mithril-relay/src/p2p/peer.rs:199]: Done :heavy_check_mark:

File: mithril-relay/src/commands/passive.rs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        let dial_to = self.dial_to.to_owned();
+        let addr: Multiaddr = format!("/ip4/0.0.0.0/tcp/{}", self.listen_port).parse()?;
+
+        let mut relay = PassiveRelay::new(&addr).start().await?;
+        if let Some(dial_to_address) = dial_to {
+            relay.dial_peer(dial_to_address.clone())?;
+        }
+        loop {
+            if let Err(err) = relay.tick().await {
+                error!("P2PClient: tick error"; "error" => format!("{err:#?}"));
```
[the remark this answers is not in the document]
Comment (jpraynaud) [reply mithril-relay/src/commands/passive.rs:32]: No, we want the relay to keep running when an error occurs :+1:

**Pull request merged**